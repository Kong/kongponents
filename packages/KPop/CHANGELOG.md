# @kongponents/kpop

## Table of Contents

- [0.2.7](#027)
- [0.2.8](#028)
- [0.2.6](#026)

### 0.2.8
 - Adds title slot & placement prop validator [#123](https://github.com/Kong/kongponents/pull/123)

### 0.2.7
 - Makes the pop-over shows in IE [#124](https://github.com/Kong/kongponents/pull/124)

### 0.2.6
 - Adds slot to title + theme overrides [#97](https://github.com/Kong/kongponents/pull/97)
