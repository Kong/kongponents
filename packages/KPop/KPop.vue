<template>
  <component :is="tag">
    <slot/>
    <div
      v-if="isSvg"
      :name="popoverTransitions">
      <foreignObject>
        <div
          v-show="isShow"
          ref="popper"
          :style="popoverStyle"
          :class="popoverClasses">
          <div
            v-if="title"
            class="k-popover-title">
            <slot name="title">{{ title }}</slot>
          </div>
          <div class="k-popover-content">
            <slot name="content"/>
          </div>
        </div>
      </foreignObject>
    </div>
    <transition
      v-else
      name="fade">
      <div
        v-show="isShow"
        ref="popper"
        :style="popoverStyle"
        :class="popoverClasses">
        <div
          v-if="title"
          class="k-popover-title">{{ title }}</div>
        <div class="k-popover-content">
          <slot name="content"/>
        </div>
      </div>
    </transition>
  </component>
</template>
<script>
import Popper from 'popper.js'

export default {
  name: 'KPop',
  props: {
    /**
     * The target element to append the popper to
     */
    target: {
      type: String,
      default: 'body'
    },
    /**
     * The tag to wrap the popover around
     */
    tag: {
      type: String,
      default: 'div'
    },
    /**
     * The title of the Popover header
     */
    title: {
      type: String,
      default: ''
    },
    /**
     * The position of the popover
     * 'top' | 'bottom' | 'left' | 'right'
     */
    placement: {
      type: String,
      default: 'top'
    },
    /**
     * How the Popover will trigger
     * 'click' | 'hover'
     */
    trigger: {
      type: String,
      default: 'click',
      validator: function (value) {
        return [
          'click',
          'hover'
        ].indexOf(value) !== -1
      }
    },
    /**
     * The width of the Popover body
     */
    width: {
      type: String,
      default: '200'
    },
    /**
     * Custom classes that will be applied to the popover
     */
    popoverClasses: {
      type: String,
      default: 'k-popover'
    },
    /**
     * Custom transition names that will be applied to the popover
     */
    popoverTransitions: {
      type: String,
      default: 'fade'
    },
    /**
     * An optional flag passed in to trigger the Popover to hide - useful for external events like zooming or panning
     */
    hidePopover: {
      type: Boolean,
      default: false
    },
    /**
     * A flag for disabling the popover
     */
    disabled: {
      type: Boolean,
      default: false
    },
    /**
    * A flag indicating whether or not the element in the slot will be an SVG element
    */
    isSvg: {
      type: Boolean,
      default: false
    }
  },

  data () {
    return {
      popper: null,
      reference: null,
      isShow: false
    }
  },

  computed: {
    popoverStyle: function () {
      return {
        width: this.width + 'px'
      }
    }
  },

  watch: {
    hidePopover: function () {
      // whenever this prop gets updated, hide the popper
      if (this.isShow) {
        this.hidePopper()
      }
    }
  },

  mounted () {
    if (this.disabled) return
    this.reference = this.$el.children[0]
    this.bindEvents()
  },

  beforeDestroy () {
    const popper = this.$refs.popper
    if (popper && this.trigger === 'click') {
      this.reference.removeEventListener('click', this.handleClick)
      popper.removeEventListener('click', this.showPopper)
      document.documentElement.removeEventListener('click', this.handleClick)
    } else if (this.reference) {
      this.reference.removeEventListener('mouseenter', this.createInstance)
      this.reference.removeEventListener('mouseleave', this.toggle)
    }

    this.destroy()
  },

  methods: {
    hidePopper () {
      if (this.trigger !== 'hover') {
        this.isShow = false
      }

      this.timer = setTimeout(() => {
        this.isShow = false
        this.destroy()
      }, 300)
    },

    showPopper () {
      this.isShow = true
      if (this.timer) clearTimeout(this.timer)
      if (this.popperTimer) clearTimeout(this.popperTimer)
    },

    async createInstance () {
      this.showPopper()
      // destroy any previous poppers before creating new one
      this.destroy()
      const placementMapper = {
        top: 'top',
        left: 'left',
        right: 'right',
        bottom: 'bottom'
      }
      const placement = placementMapper[this.placement] ? placementMapper[this.placement] : 'bottom'
      const popperEl = this.$refs.popper

      document.querySelector(this.target).appendChild(popperEl)
      await this.$nextTick()
      this.popper = new Popper(this.reference, popperEl, { placement, removeOnDestroy: true })

      await this.$nextTick()
      this.popper.update()
    },

    handleClick (e) {
      e.stopPropagation()
      if (this.reference && this.reference.contains(e.target)) {
        if (this.isShow) {
          this.hidePopper()
        } else {
          this.createInstance()
        }
      } else if (this.$refs.popper && this.$refs.popper.contains(e.target)) {
        this.showPopper()
      } else {
        if (this.isShow) this.hidePopper()
      }
    },

    bindEvents () {
      const popper = this.$refs.popper
      if (popper) {
        if (this.trigger === 'hover') {
          this.reference.addEventListener('mouseenter', this.createInstance)
          this.reference.addEventListener('mouseleave', this.hidePopper)
          popper.addEventListener('mouseenter', this.showPopper)
          popper.addEventListener('mouseleave', this.hidePopper)
        } else {
          this.reference.addEventListener('click', this.handleClick)
          popper.addEventListener('click', this.showPopper)
          document.documentElement.addEventListener('click', this.handleClick)
        }
      }
    },

    destroy () {
      if (this.popper) {
        this.isShow = false
        this.popper.destroy()
        this.popper = null
      }
    }

  }
}
</script>
<style lang="scss" scoped>
@import '~@kongponents/styles/_variables.scss';

.k-popover {
  z-index: 1000;
  max-width: none;
  font-size: var(--KPopBodySize, var(--type-sm, type(sm)));
  text-align: left;
  white-space: normal;
  color: var(--KPopColor, var(--tblack-70, color(tblack-70)));
  background-color: var(--KPopBackground, var(--twhite-1, color(twhite-1)));
  border: 1px solid var(--KPopBorder, var(--grey-84, color(grey-84)));
  border-radius: 3px;
  -webkit-box-shadow: 0 0 12px rgba(0,0,0,.12);
  box-shadow: 0 0 12px rgba(0,0,0,.12);
  padding: 1rem;

  .k-popover-title {
    padding-bottom: 1rem;
    margin-bottom: 1rem;
    font-size: var(--KPopHeaderSize, var(--type-md, type(md)));
    font-weight: 500;
    border-bottom: 1px solid rgba(0,0,0,.10);
    background-color: var(--KPopBackground, var(--twhite-1, color(twhite-1)));
  }

  &[x-placement^="bottom"] {
    margin-top: var(--spacing-md, spacing(md));

    &:after, &:before {
      bottom: 100%;
      left: 50%;
      border: solid transparent;
      content: " ";
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
    }

    &:after {
      border-color: rgba(255, 255, 255, 0);
      border-bottom-color: var(--KPopBackground, var(--twhite-1, color(twhite-1)));
      border-width: 10px;
      margin-left: -10px;
    }

    &:before {
      border-color: rgba(250, 250, 250, 0);
      border-bottom-color: var(--KPopBorder, var(--grey-84, color(grey-84)));
      border-width: 11px;
      margin-left: -11px;
    }
  }

  &[x-placement^="top"] {
    margin-bottom: var(--spacing-md, spacing(md));

    &:after, &:before {
      top: 100%;
      left: 50%;
      border: solid transparent;
      content: " ";
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
    }

    &:after {
      border-color: rgba(255, 255, 255, 0);
      border-top-color: var(--KPopBackground, var(--twhite-1, color(twhite-1)));
      border-width: 10px;
      margin-left: -10px;
    }

    &:before {
      border-color: rgba(250, 250, 250, 0);
      border-top-color: var(--KPopBorder, var(--grey-84, color(grey-84)));
      border-width: 11px;
      margin-left: -11px;
    }
  }

  &[x-placement^="left"] {
    margin-right: var(--spacing-md, spacing(md));

    &:after, &:before {
      left: 100%;
      top: 50%;
      border: solid transparent;
      content: " ";
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
    }

    &:after {
      border-color: rgba(255, 255, 255, 0);
      border-left-color: var(--KPopBackground, var(--twhite-1, color(twhite-1)));
      border-width: 10px;
      margin-top: -10px;
    }

    &:before {
      border-color: rgba(250, 250, 250, 0);
      border-left-color: var(--KPopBorder, var(--grey-84, color(grey-84)));
      border-width: 11px;
      margin-top: -11px;
    }
  }

  &[x-placement^="right"] {
    margin-left: var(--spacing-md, spacing(md));

    &:after, &:before {
      right: 100%;
      top: 50%;
      border: solid transparent;
      content: " ";
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
    }

    &:after {
      border-color: rgba(255, 255, 255, 0);
      border-right-color: var(--KPopBackground, var(--twhite-1, color(twhite-1)));
      border-width: 10px;
      margin-top: -10px;
    }

    &:before {
      border-color: rgba(250, 250, 250, 0);
      border-right-color: var(--KPopBorder, var(--grey-84, color(grey-84)));
      border-width: 11px;
      margin-top: -11px;
    }
  }
}

/** Animations */
$transition-speed: .3s !default;

@keyframes fadeIn {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
@keyframes fadeOut {
  from {
    opacity: 1;
  }
  to {
    opacity: 0;
  }
}

.fade-enter-active,
.fadeIn,
.fade-leave-active,
.fadeOut {
  animation-duration: $transition-speed;
  animation-fill-mode: both;
}
.fade-enter-active,
.fadeIn {
  animation-name: fadeIn;
}
.fade-leave-active,
.fadeOut {
  animation-name: fadeOut;
}
</style>
