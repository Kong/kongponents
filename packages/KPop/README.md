```js
const attributes = {
  width: 350,
  title: 'Cool Beans!'
  content: 'I have fantastic content!'
}
<KPop 
  :width="attributes.width"
  :title="attributes.title"
  :content="attributes.content"
>
  Click on me to see popover
</KPop>
```
