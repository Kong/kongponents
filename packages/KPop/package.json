{
  "name": "@kongponents/kpop",
  "version": "0.2.5",
  "description": "A Popover component",
  "main": "KPop.vue",
  "scripts": {
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "dependencies": {
    "popper.js": "^1.15.0"
  },
  "peerDependencies": {
    "@kongponents/styles": "0.0.2"
  }
}
