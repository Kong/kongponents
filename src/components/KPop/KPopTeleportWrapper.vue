<template>
  <template v-if="useTeleport">
    <Teleport :to="target">
      <slot />
    </Teleport>
  </template>
  <template v-else>
    <slot />
  </template>
</template>

<script setup lang="ts">
import type { PopTeleportWrapperProps, PopTeleportWrapperSlots } from '@/types'

const {
  useTeleport,
  target = null,
} = defineProps<PopTeleportWrapperProps>()

defineSlots<PopTeleportWrapperSlots>()
</script>
