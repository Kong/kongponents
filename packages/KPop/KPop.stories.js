import { storiesOf } from '../../storybook/node_modules/@storybook/vue'

import KPop from './KPop.vue'

storiesOf('KPop', module)
  .add('Popover with default positioning on top, triggers on click by default', () => ({
    components: { KPop },
    template: `<KPop title="Cool header" style="margin-top:150px;margin-left:100px;"><button>button</button><div slot="content">I am some sample text with default positioning!</div></KPop>`
  }), {info: {}})
  .add('Popover with positioning on the right', () => ({
    components: { KPop },
    template: `<KPop placement="right" title="Cool header" style="margin-top:100px;"><button>button</button><div slot="content">I am some sample text with right positioning!</div></KPop>`
  }), {info: {}})
  .add('Popover that triggers on hover', () => ({
    components: { KPop },
    template: `<KPop trigger="hover" title="Cool header" style="margin-left:100px;"><button>button</button><div slot="content">I am some sample text that triggers on hover!</div></KPop>`
  }), {info: {}})
