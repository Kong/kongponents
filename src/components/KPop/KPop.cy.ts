import { mount } from 'cypress/vue'
import KPop from '@/components/KPop/KPop.vue'
import { h } from 'vue'

describe('KPop', () => {
  it('renders props when passed', () => {
    const popButtonText = 'Click Me!'
    const popTitle = 'Cool Beans!'

    mount(KPop, {
      props: {
        buttonText: popButtonText,
        title: popTitle,
      },
    })

    cy.get('[data-testid="popover-button"]').should('contain.html', popButtonText)
    cy.get('.popover-title').should('contain.html', popTitle)
  })

  it('renders slots when passed', () => {
    const popTitle = 'Look Mah!'
    const popContent = 'Pop Content'
    const popFooter = 'Pop Footer'

    mount(KPop, {
      props: {
      },
      slots: {
        title: () => h('span', {}, popTitle),
        content: h('span', {}, popContent),
        footer: h('span', {}, popFooter),
      },
    })

    cy.get('.popover-title').should('contain.html', popTitle)
    cy.get('.popover-content').should('contain.html', popContent)
    cy.get('.popover-footer').should('contain.html', popFooter)
  })

  it('renders with correct px width', () => {
    const width = 300
    mount(KPop, {
      props: {
        width: width + '',
      },
      slots: {
        default: () => h('div', { class: ['slottedEl'] }, 'Slotted element'),
      },
    })
    cy.get('.slottedEl').click()
    cy.get('.popover .popover-container').invoke('width').should('eq', width)
  })

  it('renders with correct title', () => {
    const title = 'Cool Beans!'
    mount(KPop, {
      props: {
        title,
      },
    })
    cy.get('.popover-title').should('have.text', title)
  })

  it('has no title if no prop', () => {
    mount(KPop, {
      props: {
      },
    })
    cy.get('.popover-title').should('not.exist')
  })

  it('shows element on click', () => {
    mount(KPop, {
      props: {
        title: 'Popover Title',
      },
      slots: {
        default: () => h('div', { class: ['slottedEl'] }, 'Slotted element'),
      },
    })

    cy.get('.popover').should('not.be.visible')
    cy.get('.slottedEl').click()
    cy.get('.popover').should('be.visible')
  })

  it('shows element on hover', () => {
    mount(KPop, {
      props: {
        title: 'Popover Title',
        trigger: 'hover',
      },
      slots: {
        default: () => h('div', { class: ['slottedEl'] }, 'Slotted element'),
      },
    })

    cy.get('.popover').should('not.be.visible')
    cy.get('.slottedEl').trigger('mouseenter')
    cy.get('.popover').should('be.visible')
  })

  it('renders with correct default zIndex', () => {
    mount(KPop)

    cy.get('.popover').should('have.css', 'z-index', '1000')
  })

  it('renders with custom zIndex', () => {
    mount(KPop, {
      props: {
        zIndex: 2200,
      },
    })

    cy.get('.popover').should('have.css', 'z-index', '2200')
  })

  it('does not render close icon when prop is false', () => {
    mount(KPop, {
      props: {
        hideCloseIcon: true,
      },
    })

    cy.get('.popover-close-button').should('not.exist')
  })

  it('closes popover when close button is clicked', () => {
    mount(KPop, {
      props: {
        title: 'Popover Title',
      },
      slots: {
        default: () => h('div', { class: ['slottedEl'] }, 'Slotted element'),
      },
    })

    cy.get('.slottedEl').click()
    cy.get('.popover').should('be.visible')
    cy.get('.popover-close-button').click()
    cy.get('.popover').should('not.be.visible')
  })
})
