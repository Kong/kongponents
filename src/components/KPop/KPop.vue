<template>
  <component
    :is="tag"
    ref="kPopoverElement"
    class="k-popover"
    @keydown.esc="onEscapePress"
  >
    <div
      ref="triggerWrapperElement"
      class="popover-trigger-wrapper"
    >
      <slot>
        <KButton
          :aria-controls="popoverId"
          data-testid="popover-button"
        >
          {{ buttonText }}
        </KButton>
      </slot>
    </div>

    <KPopTeleportWrapper
      :target="target"
      :use-teleport="!!target"
    >
      <Transition name="kongponents-fade-transition">
        <div
          v-show="isVisible"
          :id="popoverId"
          ref="popoverElement"
          :aria-labelledby="$slots.title || title ? titleId : undefined"
          class="popover"
          :class="popoverClassesObj"
          role="dialog"
          :style="popoverStyles"
          :x-placement="calculatedPlacement"
        >
          <div
            class="popover-container"
            :style="popoverContainerStyles"
          >
            <button
              v-if="!hideCloseIcon"
              ref="popoverCloseButton"
              aria-label="Close popover"
              class="popover-close-button"
              :tabindex="isVisible ? 0 : -1"
              type="button"
              @click="hidePopover"
            >
              <CloseIcon
                class="popover-close-icon"
                decorative
                :size="KUI_ICON_SIZE_30"
              />
            </button>
            <div
              v-if="$slots.title || title"
              class="popover-header"
            >
              <div
                v-if="$slots.title || title"
                :id="titleId"
                class="popover-title"
                :class="{ 'close-icon-spacing': !hideCloseIcon }"
              >
                <slot name="title">
                  {{ title }}
                </slot>
              </div>
            </div>
            <div
              class="popover-content"
              :class="{ 'close-icon-spacing': !hideCloseIcon && !($slots.title || title) }"
            >
              <slot name="content" />
            </div>
            <div
              v-if="$slots.footer"
              class="popover-footer"
            >
              <slot name="footer" />
            </div>
          </div>
        </div>
      </Transition>
    </KPopTeleportWrapper>
  </component>
</template>

<script setup lang="ts">
import { ref, onMounted, onBeforeUnmount, computed, watch, useId } from 'vue'
import type { PropType } from 'vue'
import { useFloating, autoUpdate, autoPlacement, flip, shift } from '@floating-ui/vue'
import type { PopPlacements, PopTrigger } from '@/types'
import { PopPlacementsArray, PopTriggerArray } from '@/types'
import KButton from '@/components/KButton/KButton.vue'
import useUtilities from '@/composables/useUtilities'
import { CloseIcon } from '@kong/icons'
import { KUI_ICON_SIZE_30, KUI_SPACE_60 } from '@kong/design-tokens'
import KPopTeleportWrapper from './KPopTeleportWrapper.vue'
import { useEventListener } from '@vueuse/core'

const props = defineProps({
  buttonText: {
    type: String,
    default: '',
  },
  title: {
    type: String,
    default: '',
  },
  placement: {
    type: String as PropType<PopPlacements>,
    validator: (value: PopPlacements): boolean => PopPlacementsArray.includes(value),
    default: 'auto',
  },
  trigger: {
    type: String as PropType<PopTrigger>,
    default: 'click',
    validator: (value: PopTrigger): boolean => PopTriggerArray.includes(value),
  },
  popoverTimeout: {
    type: Number,
    default: 300,
  },
  hideCloseIcon: {
    type: Boolean,
    default: false,
  },
  hideCaret: {
    type: Boolean,
    default: false,
  },
  closeOnPopoverClick: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  width: {
    type: String,
    default: '200',
  },
  maxWidth: {
    type: String,
    default: 'auto',
  },
  maxHeight: {
    type: String,
    default: 'auto',
  },
  popoverClasses: {
    type: String,
    default: '',
  },
  tag: {
    type: String,
    default: 'div',
  },
  zIndex: {
    type: Number,
    default: 1000,
  },
  offset: {
    type: String,
    default: KUI_SPACE_60,
  },
  target: {
    type: [String, null],
    default: null,
  },
})

const emit = defineEmits(['open', 'close', 'popover-click'])

const { getSizeFromString } = useUtilities()

const popoverId = useId()
const titleId = useId()
const kPopoverElement = ref<HTMLElement | null>(null)
const triggerWrapperElement = ref<HTMLElement | null>(null)
const popoverElement = ref<HTMLElement | null>(null)
const isVisible = ref<boolean>(false)

const popoverTrigger = computed((): HTMLElement | null => triggerWrapperElement.value && triggerWrapperElement.value?.children[0] ? triggerWrapperElement.value?.children[0] as HTMLElement : null)

const timer = ref<number | null>(null)

const togglePopover = () => {
  if (!isVisible.value) {
    showPopover()
  } else {
    hidePopover()
  }
}

const cancelFloatingUpdates = () => {
  if (floatingUpdates.value) {
    floatingUpdates.value()
  }
}

const startFloatingUpdates = () => {
  cancelFloatingUpdates()
  if (popoverTrigger.value && popoverElement.value) {
    // start the auto updates for the popover position
    // autoUpdate cleanup function
    // docs: https://floating-ui.com/docs/autoUpdate#usage
    floatingUpdates.value = autoUpdate(popoverTrigger.value, popoverElement.value, updatePosition)
  }
}

const showPopover = async () => {
  if (!props.disabled) {
    if (timer.value) {
      clearTimeout(timer.value)
    }

    startFloatingUpdates()
    isVisible.value = true
  }
}

const hidePopover = () => {
  timer.value = setTimeout(() => {
    cancelFloatingUpdates()
    isVisible.value = false
  }, props.trigger === 'hover' ? props.popoverTimeout : 0)
}

const clickHandler = (event: Event) => {
  const target = event.target as HTMLElement

  if (popoverTrigger.value?.contains(target) && !popoverElement.value?.contains(target)) {
    // toggle popover if clicked within the trigger

    togglePopover()
  } else if (popoverElement.value?.contains(target) && !triggerWrapperElement.value?.contains(target)) {
    // emit popover-click event if clicked within the popover
    // also close the popover if closeOnPopoverClick is true

    if (props.closeOnPopoverClick) {
      hidePopover()
    }

    emit('popover-click')
  } else if (isVisible.value && !kPopoverElement.value?.contains(target)) {
    // close popover if clicked outside of the popover

    hidePopover()
  }
}

/**
 * Backwards compatibility for the placement prop
 * Converts the placement prop to the correct format for Floating UI
 * E.g.: 'topStart' -> 'top-start'
 */
const popoverPlacement = computed((): PopPlacements => props.placement.trim().replace(/ /g, '-').replace(/[A-Z]+(?![a-z])|[A-Z]/g, ($, ofs) => (ofs ? '-' : '') + $.toLowerCase()).replace(/--+/g, '-').replace(/-+$/g, '') as PopPlacements)

const { floatingStyles, placement: calculatedPlacement, update: updatePosition } = useFloating(popoverTrigger, popoverElement, {
  ...(popoverPlacement.value === 'auto' && { middleware: [autoPlacement()] }), // when placement is auto just use autoPlacement middleware
  ...(popoverPlacement.value !== 'auto' && {
    placement: popoverPlacement.value,
    middleware: [
      shift(), // Shifts the floating element to keep it in view.
      flip(), // Changes the placement of the floating element to keep it in view.
    ],
  }),
  strategy: 'fixed',
  transform: false,
})

const popoverOffset = computed(() => getSizeFromString(props.offset))
const marginStyles = computed(() => {
  if (calculatedPlacement.value.includes('top')) {
    return {
      marginBottom: popoverOffset.value,
    }
  } else if (calculatedPlacement.value.includes('bottom')) {
    return {
      marginTop: popoverOffset.value,
    }
  } else if (calculatedPlacement.value.includes('right')) {
    return {
      marginLeft: popoverOffset.value,
    }
  } else if (calculatedPlacement.value.includes('left')) {
    return {
      marginRight: popoverOffset.value,
    }
  } else {
    return {}
  }
})

const popoverStyles = computed(() => {
  return {
    ...floatingStyles.value,
    zIndex: props.zIndex,
  }
})

const popoverContainerStyles = computed(() => {
  return {
    ...marginStyles.value,
    width: getSizeFromString(props.width),
    maxWidth: getSizeFromString(props.maxWidth),
    maxHeight: getSizeFromString(props.maxHeight),
  }
})

const popoverClassesObj = computed(() => [props.popoverClasses, { 'hide-caret': props.hideCaret }])

const floatingUpdates = ref<() => void>()

const onEscapePress = (event: Event) => {
  hidePopover()
  event.stopPropagation()
}

defineExpose({
  hidePopover,
  showPopover,
})

onMounted(() => {
  if (document) {
    // handle various click events to determine how to handle the click event in a generic clickHandler function
    // we don't set any other click event listeners on purpose to avoid conflict of event listeners
    useEventListener(document, 'click', clickHandler)

    if (popoverTrigger.value) {
      // determine the element to bind aria-controls attribute to
      // data-dropdown-trigger is used to determine the default (not slotted) KDropdown trigger
      const ariaControlsElement = popoverTrigger.value.querySelector('button[data-dropdown-trigger="true"]') || popoverTrigger.value
      if (!ariaControlsElement.hasAttribute('aria-controls')) {
        ariaControlsElement.setAttribute('aria-controls', popoverId)
      }

      if (props.trigger === 'hover') {
        useEventListener(popoverTrigger.value, 'mouseenter', showPopover)
        useEventListener(popoverTrigger.value, 'focus', showPopover)
        useEventListener(popoverTrigger.value, 'mouseleave', hidePopover)
        useEventListener(popoverTrigger.value, 'blur', hidePopover)
      }
    }

    if (popoverElement.value && props.trigger === 'hover') {
      useEventListener(popoverElement.value, 'mouseenter', showPopover)
      useEventListener(popoverElement.value, 'focusin', showPopover)
      useEventListener(popoverElement.value, 'mouseleave', hidePopover)
      useEventListener(popoverElement.value, 'focusout', hidePopover)
    }
  }
})

onBeforeUnmount(() => {
  cancelFloatingUpdates()
})

watch(isVisible, (val) => {
  if (val) {
    emit('open')
  } else {
    emit('close')
  }
})
</script>

<style lang="scss" scoped>
/* Component variables */

$kPopCaretSize: 10px;
$kPopCaretShadowElementSize: 11px;
$kPopCaretOffset: 16px;

/* Component mixins */

@mixin kPopCaret {
  &:after, &:before {
    border: solid var(--kui-color-border-transparent, $kui-color-border-transparent);
    content: " ";
    height: 0;
    pointer-events: none;
    position: absolute;
    width: 0;
  }

  &:after {
    border-width: $kPopCaretSize;
    margin-left: calc($kPopCaretSize * -1);
  }

  &:before {
    border-width: $kPopCaretShadowElementSize;
    margin-left: calc($kPopCaretShadowElementSize * -1);
  }
}

/* Component styles */

.k-popover {
  display: flex;

  .popover-trigger-wrapper {
    display: inline-flex;
    width: 100%;
  }
}

// need to have these styles not nested under .k-popover so that they still apply when the popover is teleported
.popover {
  // need max-width: 100vw; and width: max-content; for Floating UI to work properly
  // gets overwritten by the size middleware once the popover is positioned
  max-width: 100vw;
  width: max-content;

  // need to wrap popover content in a container because we cannot set position: relative; as that will break the floating-ui positioning
  .popover-container {
    background-color: var(--kui-color-background, $kui-color-background);
    border: var(--kui-border-width-10, $kui-border-width-10) solid var(--kui-color-border, $kui-color-border);
    border-radius: var(--kui-border-radius-30, $kui-border-radius-30);
    box-shadow: var(--kui-shadow, $kui-shadow);
    display: flex;
    flex-direction: column;
    font-family: var(--kui-font-family-text, $kui-font-family-text);
    gap: var(--kui-space-40, $kui-space-40);
    padding: var(--kui-space-60, $kui-space-60);
    position: relative;
    text-align: left;
    white-space: normal;

    .popover-close-button {
      @include defaultButtonReset;

      // fixing mixed-decls deprecation: https://sass-lang.com/d/mixed-decls
      // stylelint-disable-next-line no-duplicate-selectors
      & {
        border-radius: var(--kui-border-radius-20, $kui-border-radius-20);
        color: var(--kui-color-text-neutral, $kui-color-text-neutral);
        margin: var(--kui-space-60, $kui-space-60) var(--kui-space-60, $kui-space-60) var(--kui-space-0, $kui-space-0) var(--kui-space-0, $kui-space-0);
        outline: none;
        position: absolute;
        right: 0;
        top: 0;
      }

      &:hover, &:focus {
        color: var(--kui-color-text-neutral-strong, $kui-color-text-neutral-strong) !important;
      }

      &:focus-visible {
        box-shadow: var(--kui-shadow-focus, $kui-shadow-focus);
      }

      .popover-close-icon {
        pointer-events: none;
      }
    }

    .popover-header {
      align-items: baseline;
      display: flex;

      .popover-title {
        color: var(--kui-color-text, $kui-color-text);
        font-size: var(--kui-font-size-40, $kui-font-size-40);
        font-weight: var(--kui-font-weight-bold, $kui-font-weight-bold);
        line-height: var(--kui-line-height-30, $kui-line-height-30);

        &.close-icon-spacing {
          margin-right: var(--kui-space-60, $kui-space-60);
        }
      }
    }

    .popover-content {
      color: var(--kui-color-text-neutral-stronger, $kui-color-text-neutral-stronger);
      font-size: var(--kui-font-size-20, $kui-font-size-20);
      font-weight: var(--kui-font-weight-regular, $kui-font-weight-regular);
      line-height: var(--kui-line-height-20, $kui-line-height-20);

      &.close-icon-spacing {
        margin-right: var(--kui-space-60, $kui-space-60);
      }
    }

    .popover-footer {
      align-items: center;
      display: flex;
      gap: var(--kui-space-40, $kui-space-40);
    }
  }

  // placement and caret styles

  &[x-placement^="top"] .popover-container {
    @include kPopCaret;

    &:after, &:before {
      left: 50%;
      top: 100%;
    }

    &:after {
      /* stylelint-disable-next-line @kong/design-tokens/use-proper-token */
      border-top-color: var(--kui-color-background, $kui-color-background);
    }

    &:before {
      border-top-color: var(--kui-color-border, $kui-color-border);
    }
  }

  &[x-placement^="right"] .popover-container {
    @include kPopCaret;

    &:after, &:before {
      right: 100%;
      top: 50%;
      transform: translateY(-50%);
    }

    &:after {
      /* stylelint-disable-next-line @kong/design-tokens/use-proper-token */
      border-right-color: var(--kui-color-background, $kui-color-background);
    }

    &:before {
      border-right-color: var(--kui-color-border, $kui-color-border);
    }
  }

  &[x-placement^="bottom"] .popover-container {
    @include kPopCaret;

    &:after, &:before {
      bottom: 100%;
      left: 50%;
    }

    &:after {
      /* stylelint-disable-next-line @kong/design-tokens/use-proper-token */
      border-bottom-color: var(--kui-color-background, $kui-color-background);
    }

    &:before {
      border-bottom-color: var(--kui-color-border, $kui-color-border);
    }
  }

  &[x-placement^="left"] .popover-container {
    @include kPopCaret;

    &:after, &:before {
      left: 100%;
      top: 50%;
      transform: translate(50%, -50%);
    }

    &:after {
      /* stylelint-disable-next-line @kong/design-tokens/use-proper-token */
      border-left-color: var(--kui-color-background, $kui-color-background);
    }

    &:before {
      border-left-color: var(--kui-color-border, $kui-color-border);
    }
  }

  &[x-placement^="top-start"] .popover-container,
  &[x-placement^="bottom-start"] .popover-container {
    &:after, &:before {
      left: $kPopCaretOffset;
    }
  }

  &[x-placement^="right-start"] .popover-container,
  &[x-placement^="left-start"] .popover-container {
    &:after, &:before {
      top: $kPopCaretOffset;
    }
  }

  &[x-placement^="top-end"] .popover-container,
  &[x-placement^="bottom-end"] .popover-container {
    &:after, &:before {
      left: calc(100% - $kPopCaretOffset);
    }
  }

  &[x-placement^="right-end"] .popover-container,
  &[x-placement^="left-end"] .popover-container {
    &:after, &:before {
      top: calc(100% - $kPopCaretOffset);
    }
  }

  &.hide-caret .popover-container {
    &:after,
    &:before {
      display: none;
    }
  }
}
</style>
