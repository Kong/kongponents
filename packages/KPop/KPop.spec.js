import { mount } from '@vue/test-utils'
import KPop from '@/KPop/KPop'

describe('KPop', () => {
  it('matches snapshot', () => {
    const wrapper = mount(KPop, {
      propsData: {
        'width': '350',
        'title': 'Cool Beans!',
        'content': `I am a cool popover!`
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with correct width', () => {
    const wrapper = mount(KPop, {
      propsData: {
        'width': '350'
      }
    })
    const popper = wrapper.find('.k-popover')

    expect(popper.element.style['width']).toEqual('350px')
  })

  it('renders with correct title', () => {
    const wrapper = mount(KPop, {
      propsData: {
        'title': 'Cool Beans!'
      }
    })
    const popperTitle = wrapper.find('.k-popover-title')

    expect(popperTitle.text()).toEqual('Cool Beans!')
  })
})
