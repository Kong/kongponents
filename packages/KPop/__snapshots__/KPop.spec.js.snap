// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KPop matches snapshot 1`] = `
<div content="I am a cool popover!">
  <div class="k-popover" style="width: 350px; display: none;" name="fade">
    <div class="k-popover-title">Cool Beans!</div>
    <div class="k-popover-content"></div>
  </div>
</div>
`;
